### PR TITLE
Encoding Interpretation Groups

### PR DESCRIPTION
**Pull request by mobiusklein**
This PR is an (almost) complete implementation of Interpretations as defined in section section `3.5.1.[4,5,7]`. It is currently missing the `InterpretationMember` section (`3.5.1.6`), but I'm working on it.

## Adds `InterpretationCollection`
A mapping from identifier to `Interpretation`. Necessary to represent a spectrum with multiple interpretations.

## Adds `Interpretation`
A mapping from identifier to `Analyte`, also containing an `AttributeManager` to hold CV attributes describing the whole interpretation. Will eventually host another subsidiary mapping from identifier to `InterpretationMember`. A very busy object.

## Changes to `Spectrum`
`Spectrum.interpretations` will return an `InterpretationCollection`, letting the user access each `Interpretation` by its identifier. `Spectrum.analytes` remains, providing a proxy mapping projecting a mapping over all `Analyte` objects in all `Interpretation` objects in `Spectrum.interpretations`.

## Changes to Backends
The Text and JSON backends now parse `Interpretation` sections and organize `Analyte` sections under the most recent `Interpretation`. If an `Analyte` section is encountered without an `Interpretation`, it will automatically be added to a new empty `Interpretation`.  The writing behavior now includes `Interpretation` sections accordingly.

## Other Changes
I tried to apply type annotations to as much of the public interface as possible.

### CODE REVIEW

**Comment by mobiusklein** [comment]
After polishing a bit of the work I had lingering unmerged, I compared my implementation to the exact text of section `3.5.1.5`, and it doesn't match what I remembered, and section `3.5.1.5` doesn't match the examples below in `3.5.1.7`. The text states that an `Interpretation` *MUST* have an `analyte mixture members` parameter, but the examples (which my implementation matches), instead using order of appearance:
```
<Interpretation=1>
MS:1002357|PSM-level probability=0.974
<Analyte=1>
MS:1000888|peptidoform=DSDDVPM[Ox]VLVGNKCDLAAR
<Analyte=2>
MS:1000888|peptidoform=VLVGNKCDLDSDDVPMVLR
<Interpretation=2>
MS:1002357|PSM-level probability=0.937
<Analyte=3>                                    	
MS:1000888|peptidoform=CDLDSDDSDDVPMVLSAR
```
The text would have required:
```
<Analyte=1>
MS:1000888|peptidoform=DSDDVPM[Ox]VLVGNKCDLAAR
<Analyte=2>
MS:1000888|peptidoform=VLVGNKCDLDSDDVPMVLR
<Analyte=3>                                    	
MS:1000888|peptidoform=CDLDSDDSDDVPMVLSAR
<Interpretation=1>
MS:XXXXXX|analyte mixture members=1,2
MS:1002357|PSM-level probability=0.974
<Interpretation=2>
MS:XXXXXX|analyte mixture members=3
MS:1002357|PSM-level probability=0.937
```
Does anyone remember which method we settled on?

**Comment by edeutsch** [comment]
After discussion last week, I believe we have settled on:
```
<Spectrum>
MS:1000744|selected ion m/z=880.8902
<Analyte=1>
MS:1000888|peptidoform=DSDDVPMVLVGNKCDLAAR
<Interpretation=1>
MS:1002357|PSM-level probability=0.974
```

and

```
<Spectrum>
MS:1000744|selected ion m/z=880.8902
<Analyte=1>
MS:1000888|peptidoform=DSDDVPM[Oxidation]VLVGNKCDLAAR
<Analyte=2>
MS:1000888|peptidoform=VLVGNKCDLDSDDVPMVLR
<Analyte=3>                                        
MS:1000888|peptidoform=CDLDSDDSDDVPMVLSAR
<Interpretation=1>
MS:XXXXXXX|analyte mixture members=1,2
MS:1002357|PSM-level probability=0.974
<Interpretation=2>
MS:XXXXXXX|analyte mixture members=3
MS:1002357|PSM-level probability=0.937
```

and

```
<Spectrum>
MS:1000744|selected ion m/z=880.8902
<Analyte=1>
MS:1000888|peptidoform=DSDDVPM[Oxidation]VLVGNKCDLAAR
<Analyte=2>
MS:1000888|peptidoform=VLVGNKCDLDSDDVPMVLR
<Analyte=3>                                        
MS:1000888|peptidoform=CDLDSDDSDDVPMVLSAR
<Interpretation=1>
MS:XXXXXXX|analyte mixture members=1,2
MS:1002357|PSM-level probability=0.974
<InterpretationMember=1>
MS:XXXXXXX|explained intensity fraction=0.287
<InterpretationMember=2>
MS:XXXXXXX|explained intensity fraction=0.423
<Interpretation=2>
MS:XXXXXXX|analyte mixture members=3
MS:1002357|PSM-level probability=0.937
<InterpretationMember=3>
MS:XXXXXXX|explained intensity fraction=0.512
```

Please let me know if there are any errors here.

**Comment by mobiusklein** [comment]
I think I've updated things to work with the new hierarchy. Here's a JSON version of the most complex case:
```json
    {
      "analytes": {
        "1": {
          "attributes": [
            {
              "accession": "MS:1000888",
              "name": "peptidoform",
              "value": "DSDDVPM[Oxidation]VLVGNKCDLAAR"
            }
          ],
          "id": "1"
        },
        "2": {
          "attributes": [
            {
              "accession": "MS:1000888",
              "name": "peptidoform",
              "value": "VLVGNKCDLDSDDVPMVLR"
            }
          ],
          "id": "2"
        },
        "3": {
          "attributes": [
            {
              "accession": "MS:1000888",
              "name": "peptidoform",
              "value": "CDLDSDDSDDVPMVLSAR"
            }
          ],
          "id": "3"
        }
      },
      "annotations": [],
      "attributes": [
        {
          "accession": "MS:1003061",
          "name": "spectrum name",
          "value": "Test"
        },
        {
          "accession": "MS:1000744",
          "name": "selected ion m/z",
          "value": 880.8902
        }
      ],
      "intensities": [],
      "interpretations": {
        "1": {
          "attributes": [
            {
              "accession": "MS:1002357",
              "name": "PSM-level probability",
              "value": 0.974
            },
            {
              "accession": "MS:XXXXXXX",
              "name": "analyte mixture members",
              "value": "1,2"
            }
          ],
          "id": "1",
          "members": {
            "1": {
              "attributes": [
                {
                  "accession": "MS:XXXXXXX",
                  "name": "explained intensity fraction",
                  "value": 0.287
                }
              ],
              "id": "1"
            },
            "2": {
              "attributes": [
                {
                  "accession": "MS:XXXXXXX",
                  "name": "explained intensity fraction",
                  "value": 0.423
                }
              ],
              "id": "2"
            }
          }
        },
        "2": {
          "attributes": [
            {
              "accession": "MS:1002357",
              "name": "PSM-level probability",
              "value": 0.937
            },
            {
              "accession": "MS:XXXXXXX",
              "name": "analyte mixture members",
              "value": "3"
            }
          ],
          "id": "2",
          "members": {
            "3": {
              "attributes": [
                {
                  "accession": "MS:XXXXXXX",
                  "name": "explained intensity fraction",
                  "value": 0.512
                }
              ],
              "id": "3"
            }
          }
        }
      },
      "mzs": []
    }
```
The JSON version is very heavy on blank space, making it hard to parse by eye, but the model looks right.

**Comment by mobiusklein** [comment]
This still does not enforce controlled vocabulary parameters, but the structure for the interpretations is now consistent with the specification.

File: implementations/python/mzlib/analyte.py [hunk 1]
```diff
@@ -1,20 +1,194 @@
...
 import textwrap
+from typing import Iterable, KeysView, ItemsView, ValuesView, Dict
+
+from mzlib.attributes import AttributedEntity, AttributeManager, IdentifiedAttributeManager
+
+
+FIRST_ANALYTE_KEY = '1'
+FIRST_INTERPRETATION_KEY = '1'
+
+ANALYTE_MIXTURE_TERM = "MS:XXXXXXX|analyte mixture members"
```
Comment (bittremieux) [inline implementations/python/mzlib/analyte.py:17]: Just flagging this as a reminder to add the proper CV term.

File: implementations/python/mzlib/analyte.py [hunk 2]
```diff
@@ -1,20 +1,194 @@
...
+except ImportError:
+    from collections import (MutableMapping, Mapping)
+
 import textwrap
+from typing import Iterable, KeysView, ItemsView, ValuesView, Dict
+
+from mzlib.attributes import AttributedEntity, AttributeManager, IdentifiedAttributeManager
+
+
+FIRST_ANALYTE_KEY = '1'
```
Comment (bittremieux) [inline implementations/python/mzlib/analyte.py:14]: These identifiers are explicitly defined as integers in the specification. What is the advantage of encoding them as strings here?

File: implementations/python/mzlib/analyte.py [hunk 3]
```diff
@@ -1,20 +1,194 @@
...
+
+class _AnalyteMappingProxy(Mapping):
+    def __init__(self, parent):
+        self.parent = parent
+
+    def __getitem__(self, key):
+        for group_id, group in self.parent.items():
+            try:
+                return group[key]
+            except KeyError:
```
Comment (bittremieux) [inline implementations/python/mzlib/analyte.py:28]: `if key in group` should be slightly more efficient than a try-except block.

File: implementations/python/mzlib/analyte.py [hunk 4]
```diff
@@ -1,20 +1,194 @@
...
+    def __iter__(self):
+        keys = set()
+        for group_id, group in self.parent.items():
+            k_ = set(group.keys())
+            assert not (keys & k_)
+            keys.update(k_)
+        return iter(keys)
+
+    def __len__(self):
+        n = 0
```
Comment (bittremieux) [inline implementations/python/mzlib/analyte.py:41]: `return sum([len(group) for group in self.parent.values()])`

File: implementations/python/mzlib/annotation.py [hunk 5]
```diff
@@ -149,6 +152,17 @@ class IonAnnotationBase(object, metaclass=SeriesLabelSubclassRegisteringMeta):
     series_label = None
     _molecule_description_fields = {}
 
+    series: str
+    neutral_losses: list
```
Comment (bittremieux) [inline implementations/python/mzlib/annotation.py:156]: `List`

Same for `adducts` two lines down and return values of `adduct()` and `neutral_loss()`.

File: implementations/python/mzlib/attributes.py [hunk 6]
```diff
@@ -16,10 +18,15 @@ class AttributeManager(object):
         The number of attribute groups assigned.
 
     """
+    attributes: list
```
Comment (bittremieux) [inline implementations/python/mzlib/attributes.py:21]: `List` and `Dict`

File: implementations/python/mzlib/backends/base.py [hunk 7]
```diff
@@ -330,17 +293,17 @@ def read(self):
                     break
                 yield self._parse(buffer, i)
 
-    def _get_lines_for(self, offset):
+    def _get_lines_for(self, offset: int) -> list:
```
Comment (bittremieux) [inline implementations/python/mzlib/backends/base.py:296]: `List`

File: implementations/python/mzlib/backends/base.py [hunk 8]
```diff
@@ -330,17 +293,17 @@ def read(self):
...
             spectrum_buffer = self._buffer_from_stream(infile)
             #### We will end up here if this is the last spectrum in the file
         return spectrum_buffer
 
-    def _parse(self, buffer, spectrum_index=None):
+    def _parse(self, buffer: Iterable, spectrum_index: int=None):
         raise NotImplementedError()
 
-    def search(self, specification, **query_keys):
+    def search(self, specification, **query_keys) -> list:
```
Comment (bittremieux) [inline implementations/python/mzlib/backends/base.py:306]: `List`

File: implementations/python/mzlib/backends/base.py [hunk 9]
```diff
@@ -356,21 +319,35 @@ class SpectralLibraryWriterBase(object, metaclass=SubclassRegisteringMetaclass):
...
 
+    def _filter_attributes(self, attributes: Attributed, filter_fn: Callable) -> Iterable:
+        if isinstance(attributes, AttributedEntity):
+            attributes = attributes.attributes
+        for attrib in attributes:
+            if filter_fn(attrib):
+                yield attrib
+
+    def _not_analyte_mixture_term(self, attrib):
+        if attrib:
```
Comment (bittremieux) [inline implementations/python/mzlib/backends/base.py:330]: `return not attrib or attrib[0] != ANALYTE_MIXTURE_TERM`
Doesn't really make a difference.

File: implementations/python/mzlib/backends/msp.py [hunk 10]
```diff
@@ -276,7 +279,7 @@ def create_index(self):
 
         return n_spectra
 
-    def _buffer_from_stream(self, infile):
+    def _buffer_from_stream(self, infile: io.BufferedIOBase) -> list:
```
Comment (bittremieux) [inline implementations/python/mzlib/backends/msp.py:282]: `List`

File: implementations/python/mzlib/backends/msp.py [hunk 11]
```diff
@@ -409,11 +412,14 @@ def _parse_comment(self, value, attributes):
             else:
                 attributes[item] = None
 
-    def _make_spectrum(self, peak_list, attributes):
+    def _make_spectrum(self, peak_list: list, attributes: Attributed):
```
Comment (bittremieux) [inline implementations/python/mzlib/backends/msp.py:415]: `List`

File: implementations/python/mzlib/backends/text.py [hunk 12]
```diff
@@ -191,7 +210,7 @@ def create_index(self):
 
         return n_spectra
 
-    def _buffer_from_stream(self, infile):
+    def _buffer_from_stream(self, infile: io.IOBase) -> list:
```
Comment (bittremieux) [inline implementations/python/mzlib/backends/text.py:213]: `List`

**Review by bittremieux — COMMENTED** [review]
Just some minor coding comments. At this point, the code is too complex to pick out subtle bugs by looking at it, but the examples look good. The proof of the pudding is in the eating (and the tests).

**Comment by mobiusklein** [comment]
I implemented the CV term look up and type coercion we had previously discussed as a blocking issue to remove the special treatment for the `analyte mixture members` term.

I did this by updating `psims`, which holds my CV parsing code, along with all the machinery for fetching and caching versions of the various CVs. I tried initially to add the same behavior to the included `Ontology` and `OntologyTerm` types in `mzlib` itself, but without a term graph, I would only be adding to the long list of special cases in `OntologyTerm.parse`. I thought the type was already trying to do too much to accommodate PSI-MS and UNIMOD simultaneously.

**Review by edeutsch — APPROVED** [review]
Looks great!

**Review by bittremieux — APPROVED** [review]
